### ISSUE TITLE
Editor: [code] shortcode content is changed in Calypso editor

### ISSUE THREAD

**Issue by carinapilar** [Automattic/wp-calypso#1426]
The Calypso Editor is having different behavior when using [code] to languages like XML, when you switch between the editors. If you add the content on the Visual Editor, it changes the code, and if you add on the HTML Editor looks like it's resolving the code. Here is what I did:

1 - Adding the code on the HTML Editor, then changing to the Visual Editor:

![codeoncalypso](https://cloud.githubusercontent.com/assets/16104522/11691822/f0a5605a-9e83-11e5-8483-b1c8bf2f685a.gif)

2 - Adding the code on the Visual Editor, then changing to the HTML Editor:

![codeoncalypso2](https://cloud.githubusercontent.com/assets/16104522/11691832/f9cfa730-9e83-11e5-82f7-b0f8eb74a0f4.gif)

3 - Another code, added using the HTML Editor, then changing to the Visual Editor:

![codeoncalypso3](https://cloud.githubusercontent.com/assets/16104522/11691843/00f2d226-9e84-11e5-84bc-fc473c012963.gif)

4 - Another code, added using the Visual Editor, then changing to the HTML Editor:

![codeoncalypso4](https://cloud.githubusercontent.com/assets/16104522/11691849/078f5492-9e84-11e5-823e-bfb6b1011785.gif)

**Comment by gziolo** [comment]
It also removes all spaces at the beginning of each line which is super annoying!

![removing lines](https://cloud.githubusercontent.com/assets/699132/11710777/ed4b02e0-9f21-11e5-8f89-0baf0098216c.gif)

**Comment by gziolo** [comment]
This is related to #1023.

**Comment by designsimply** [comment]
Another example was reported at https://en.forums.wordpress.com/topic/bug-symbols-in-code-blocks-converted-to-html-escape-sequences?replies=8 /hat tip ioganegambaputifonguser

Steps to reproduce:

1. Start a new post at https://wordpress.com/post
1. Switch to the HTML tab
1. Add sample code with custom attributes (see below)
1. Switch to the Visual tab then back to the HTML tab and check to see what happens to the code

<strong>Result:</strong> the code should be preserved but most of it is stripped

Sample code:

```
[code language="html"]
<div ng-repeat="entity in selector">
<label>
<input type="checkbox" ng-model="entity.val" ng-change="updateModel(entity)"></input>
<span ng-bind="entity.value"></span>
</label>
</div>
[/code]
```

After switching from Visual to HTML it becomes:

```
[code language="xml"]
<div><label>
<input type="checkbox" />

</label></div>
[/code]
```

**Comment by noam87** [comment]
Will also happen with latex `$$ &s=2` will escape the `&` into `&amp;`.

**Comment by KokkieH** [comment]
I have another example of this reported in http://en.forums.wordpress.com/topic/writing-sourcecode-in-wordpresscom with a slight difference.

In this case the user didn't switch between editors, but created and published in the visual editor. When they later reopened the post for editing, still in the visual editor, the special characters had changed into HTML entities. Closing and opening again for editing converts the ampersand in the HTML entity into an HTML entity itself, at which point the code no longer renders the correct character in the published post.

I was able to reproduce this as well using their workflow, and I was also able to reproduce by using only the text editor, so switching editors do not appear to be a factor.

**Comment by bengarrett** [comment]
I too have been suffering this problem. Basically the Calypso editor on wordpress.com converts all ampersand symbols to escaped HTML entities regardless of the context.

`&` is sanitised to `&amp;` but then eventually will be saved as `&amp;amp;` to output `&amp;`

`<p>` will be saved as `&lt;p&gt;` and then again into `&amp;lt;p&amp;gt;` to output `&lt;p&gt;`

I don't know if this issue has been fixed and the code on wordpress.com has not been updated. Or if there is a plugin conflict on wordpress.com. As the [Desktop App](https://developer.wordpress.com/calypso/) 1.2.5 handles this better but it's still not perfect.

This now works with the Desktop App in HTML view and will display as expected.
```
[code language="html"]
&lt;body>
  &lt;p>Hello & world&lt;/p>
&lt;/body>
[/code]
```
```
1 <body>
2   <p>Hello & I rock this world</p>
3 </body>
```

You edit the code in either view and the code block remains workable.

But if you paste the same code into the Visual view it still breaks just like on wordpress.com.
```
[code language="html"]
&amp;lt;body&gt;
  &amp;lt;p&gt;Hello &amp; world&amp;lt;/p&gt;
&amp;lt;/body&gt;
[/code]
```

**Comment by designsimply** [comment]
Likely culprit for the latest encoding issues mentioned by bengarrett: https://github.com/Automattic/wp-calypso/pull/1690

**Comment by vparkhere** [comment]
My issue is the same as @KokkieH, where it's a post revision that changes the symbols to their HTML notations.  Here are the steps to reproduce:

![test code 001](https://cloud.githubusercontent.com/assets/11635760/13327279/0641aa3e-db9e-11e5-8df3-ae154dfd8752.jpeg)

Follow-up: 2556485-t

**Comment by DannyvanderKraan** [comment]
I've got the exact same issue. Is there an ETA on the fix?

**Comment by NelsonMinar** [comment]
It's worth noting the HTML requoting bug has been on wordpress.com since long before Calypso was released. Here's a 4 year old post complaining about the same bug in TinyMCE: https://wordpress.org/support/topic/html-code-reformatted-by-visual-editor

**Comment by designsimply** [comment]
NelsonMinar, thank you for the note about the age of the issue. We should try to also look at severity and popularity of the requests too.

**Comment by kriskarkoski** [comment]
It looks like this has improved significantly since the last activity. Using the sample code from [here](https://github.com/Automattic/wp-calypso/issues/1426#issuecomment-164005382) pasted into the HTML editor at wordpress.com/post I see this:

<img width="700" alt="screen shot 2016-08-01 at 2 45 37 pm" src="https://cloud.githubusercontent.com/assets/1900444/17305113/fc631d6a-57f6-11e6-86bb-a8666d11b78c.png">

Switching to Visual I see this:

<img width="718" alt="screen shot 2016-08-01 at 2 45 55 pm" src="https://cloud.githubusercontent.com/assets/1900444/17305120/06409812-57f7-11e6-9aa8-c455d03caa78.png">

And back to HTML I see this (only change I see being </label></div> moved to one line):

<img width="701" alt="screen shot 2016-08-01 at 2 46 02 pm" src="https://cloud.githubusercontent.com/assets/1900444/17305137/1e72d882-57f7-11e6-998a-777b8b8bba74.png">

Would someone else who reported this mind checking to see if there's issues with other languages still (the sample specified HTML)?

I will report the code breaking out of the background highlighting in a separate issue if there isn't one already, and will also test the line-combining in WP Admin to see if it could be a TinyMCE issue.

**Comment by alisterscott** [comment]
I've been seeing this issue a lot on my own blog where I post sourcecode, particuarly ES code with the "=>" characters.

I've worked out a pattern to reproduce the issue:

1. Create a new post with source code containing =>

eg.

```
import randomlyPlaceMines from '.';
import {reverse} from 'lodash';

describe('randomlyPlaceMines', () => {
  it('should choose specified number of mines and exclude specified row and column', () => {
    const row = 0;
    const column = 0;
    const configuration = {dimensions: [1, 2], mine_count: 1};
    const mines = randomlyPlaceMines(configuration, row, column);
    assert.deepEqual(mines, [[0, 1]]);
  });
});
```

2. Save and publish code - all should be fine
3. Leave the editor - return to posts list
4. Click edit on the post - the post now loads in the rich editor but the => has been replaced with =&gt;

cc: @aduth

**Comment by azaozz** [comment]
The double-escaped chars are caused by not running the post content through the `the_editor_content` PHP filter before loading it in the editor. That filter is used in the SyntaxHighlighter Evolved plugin to html-decode the content of the shortcode: https://plugins.trac.wordpress.org/browser/syntaxhighlighter/trunk/syntaxhighlighter.php#L518.

Not sure if the filter should be run when outputting the post content for the Calypso editor as other things are probably attached to it. Maybe better to use `$SyntaxHighlighter->the_editor_content( $content )` directly.

**Comment by lancewillett** [comment]
Noting the pull request at #8325 is ready for review and will fix this issue (needs code review; UI testing is marked as successful).

**Comment by gziolo** [comment]
It should be now fixed with changes deployed with #8325. Props to @aduth! 💯
